Lacrit/assignment4/islands.cpp

```diff
@@ -0,0 +1,68 @@
+#include <stdio.h>
+#include <string.h>
+#include <stdbool.h>
+ 
+const int rows = 5;
+const int cols = 5;
+
+// A function to check if a given cell (row, col) can be included in DFS
+int isValid(int mat[rows][cols], int row, int col, bool visited[rows][cols])
+{
+    // row number is in range, column number is in range and value is 1 
+    // and not yet visited
+    return (row >= 0) && (row < rows) &&     
+           (col >= 0) && (col < cols) &&      
+           (mat[row][col] && !visited[row][col]); 
+}
+ 
+// check all 8 neighbours
+void DFS(int mat[rows][cols], int row, int col, bool visited[rows][cols])
+{
+    // These arrays are used to get row and column numbers of 8 neighbours 
+    // of a given cell
+    static int rowInd[] = {-1, -1, -1,  0, 0,  1, 1, 1};
+    static int colInd[] = {-1,  0,  1, -1, 1, -1, 0, 1};
+ 
+    // Mark this cell as visited
+    visited[row][col] = true;
+ 
+    // Recur for all connected neighbours
+    for (int k = 0; k < 8; ++k)
```

Comment:
It may be slightly easier to have two cycles:
```c++
for (int dx : {-1, 0, 1}) {
  for (int dy : {-1, 0, 1}) {
    ...
  }
}
```
or using `for (int dx = -1; dx <= 1; dx++) { ... } `

Note that isValid will return false when dx==0 & dy==0, since that cell is already painted as visited.

But since we need to iterate over just 4 neighbours is probably easier as you did :)